app/services/record_declarations/base.rb

```diff
@@ -106,9 +96,56 @@ def create_record!
       end
     end
 
+    def valid_courses_for_user
+      valid_courses = []
+      valid_courses << "ecf-mentor" if user.mentor?
+      valid_courses << "ecf-induction" if user.early_career_teacher?
+      valid_courses += NPQCourse.all.map(&:identifier) if user.npq?
```

Comment:
It's a design smell for the Base classes to have knowledge of subclasses and this is just proxying that.

Reply:
Base class has no knowledge what subclasses are doing.

What this computes is the valid courses for the user. Those do not change in child classes. We then check that the course we are sent is in the list of courses valid for the user. This again, does not change in child classes.

In fact, which chilld class is used to validate is picked on the basis of course identifier. So if we wanted to move this check to child classes, we would have to validate that the user has access to the course which the child class corresponds to. Which I guess could make sense...

Something like
```
# In mentor child class

def validate_user_can_access_course
  errors.add(...) unless user.mentor?
end
```

But definitely not in this PR. The only reason it touches this stuff is to move it from non-private to private.

Reply:
Sure. The "can access course" shouldn't be necessary if the factory is doing the correct job. This check was more relevant for schema validation, so the mentor child class should only need to validate that it's the correct course without the unless user.mentor? That would be implied by the fact that you're using that class in the first place.